Objects/unicodeobject.c

```diff
@@ -3846,10 +3870,27 @@ PyUnicode_AsUTF8(PyObject *unicode)
 {
     Py_ssize_t size;
     const char *utf8 = PyUnicode_AsUTF8AndSize(unicode, &size);
-    if (utf8 != NULL && strlen(utf8) != (size_t)size) {
-        PyErr_SetString(PyExc_ValueError, "embedded null character");
+    if (utf8 == NULL) {
         return NULL;
     }
+
+    // Cache to avoid calling O(n) strlen() operation at every
+    // PyUnicode_AsUTF8() call on the same object.
+    if (_PyUnicode_STATE(unicode).embed_null == 2) {
```

Comment:
You can make the fast path faster.
```c
    if (_PyUnicode_STATE(unicode).embed_null != 0) {
        if (_PyUnicode_STATE(unicode).embed_null == 2) {
            ...
        }
        if (_PyUnicode_STATE(unicode).embed_null == 1) {
            ...
        }
    }
    return utf8;
```
or
```c
    if (_PyUnicode_STATE(unicode).embed_null == 0) {
        return utf8;
    }
    if (_PyUnicode_STATE(unicode).embed_null == 2) {
        ...
```

Reply:
Ok, I applied your suggestion.